File: apina-core/src/main/java/fi/evident/apina/model/ClassDefinition.java
package fi.evident.apina.model;

import fi.evident.apina.model.type.ApiTypeName;

import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedHashMap;

import static java.util.Collections.unmodifiableCollection;
import static java.util.Objects.requireNonNull;

public final class ClassDefinition {

    private final ApiTypeName type;

    private final HashMap<String, PropertyDefinition> properties = new LinkedHashMap<>();

    public ClassDefinition(ApiTypeName type) {
        this.type = requireNonNull(type);
    }

    public ApiTypeName getType() {
        return type;
    }

    public Collection<PropertyDefinition> getProperties() {
        return unmodifiableCollection(properties.values());
    }

    public void addProperty(PropertyDefinition property) {
        PropertyDefinition old = properties.putIfAbsent(property.getName(), property);
        if (old != null)
            throw new IllegalArgumentException("duplicate property: " + property.getName());
    }

    public boolean hasProperty(String name) {
        return properties.containsKey(requireNonNull(name));
    }

    @Override
    public String toString() {
        return type.toString();
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/model/ClassDefinition.kt
package fi.evident.apina.model

import fi.evident.apina.model.settings.OptionalTypeMode
import fi.evident.apina.model.type.ApiType
import fi.evident.apina.model.type.ApiTypeName
import java.util.*
import java.util.Collections.unmodifiableCollection

class ClassDefinition(val type: ApiTypeName,
                      val variables: List<ApiType.Variable>) {

    private val _properties = TreeMap<String, PropertyDefinition>()
    private val _types = mutableSetOf<ApiType.Class>()

    val genericType = ApiType.Class(type, variables)

    val properties: Collection<PropertyDefinition>
        get() = _properties.values

    val types: Collection<ApiType>
        get() = unmodifiableCollection(_types)

    // FIXME: properties need to be per type
    fun addProperty(property: PropertyDefinition) {
        val old = _properties.putIfAbsent(property.name, property)
        if (old != null)
            throw IllegalArgumentException("duplicate property: ${property.name}")
    }

    fun addType(type: ApiType.Class) {
        _types.add(type)
    }

    fun hasProperty(name: String) = name in _properties

    fun toTypeScript(optionalTypeMode: OptionalTypeMode) = genericType.toTypeScript(optionalTypeMode)
    fun toSwift() = genericType.toSwift()

    override fun toString() = type.toString()
}
